raft/raft_log.cpp

```diff
@@ -0,0 +1,165 @@
+                {
+                    return (entry.entry_type == bzn::log_entry_type::single_quorum) || (entry.entry_type == bzn::log_entry_type::joint_quorum);
+                });
+        if(result == this->log_entries.crend())
+        {
+            throw std::runtime_error(MSG_NO_PEERS_IN_LOG);
+        }
+        return *result;
+    }
+
+
+    void
+    raft_log::append_log_disk(const bzn::log_entry& log_entry)
+    {
+        if (!this->log_entry_out_stream.is_open())
+        {
+            boost::filesystem::path path{this->entries_log_path};
+            if(!boost::filesystem::exists(path.parent_path()))
+            {
+                boost::filesystem::create_directories(path.parent_path());
+            }
+            this->log_entry_out_stream.open(path.string(), std::ios::out |  std::ios::binary | std::ios::app);
+        }
+        this->log_entry_out_stream << log_entry;
+        this->log_entry_out_stream.flush();
+    }
+
+
+    void
+    raft_log::leader_append_entry(const bzn::log_entry& log_entry)
+    {
+        LOG(debug) << "Appending " << log_entry_type_to_string(log_entry.entry_type) << " to my log: " << log_entry.msg.toStyledString();
+
+        this->log_entries.emplace_back(log_entry);
+        this->append_log_disk(log_entry);
+    }
+
+
+    void
+    raft_log::follower_insert_entry(size_t index, const bzn::log_entry& log_entry)
+    {
+        // case 0: the index is in the log and we agree
+        // case 1: the index is in the log and we disagree
+        // case 2: the index is right after the log
+        // case 3: the index is after the log
+
+        if (index < this->log_entries.size() &&  log_entry.term == this->log_entries[index].term)
+        {
+            // we already have accepted this entry
+            return;
+        }
+
+        if(index < this->log_entries.size() &&  log_entry.term != this->log_entries[index].term)
+        {
+            // throw away vector from index
+            this->log_entries.erase( this->log_entries.begin() + index, this->log_entries.end());
+            this->log_entries.emplace_back(log_entry);
+
+            this->rewrite_log();
+            this->append_log_disk(log_entry);
+            return;
+        }
+
+        if(index == this->log_entries.size())
+        {
+            this->log_entries.emplace_back(log_entry);
+            this->append_log_disk(log_entry);
+            return;
+        }
+
+        if(index > this->log_entries.size())
+        {
+            throw std::runtime_error(MSG_TRYING_TO_INSERT_INVALID_ENTRY);
+        }
+    }
+
+
+    bool
+    raft_log::entry_accepted(size_t previous_index, size_t previous_term) const
+    {
+        if(previous_index < this->log_entries.size()-1)
+        {
+            return false;
+        }
+        return previous_term == this->log_entries[previous_index].term;
+    }
+
+
+    void
+    raft_log::rewrite_log()
+    {
+        boost::filesystem::path path{this->entries_log_path};
```

Comment:
Could path be a member variable that's not constructed over and over?

Reply:
Good point. Turns out that boost::filesystem::remove will take a std::string.. Fixed.